### PR TITLE
Update index.md

### PR DESCRIPTION
**Pull request by jacobchughes**
Added a ZKPR section.

### CODE REVIEW

File: source/index.md [hunk 1]
```diff
@@ -1127,13 +1127,529 @@ saife.removeVolume(vol);
...
+        PrintStream pwout = new PrintStream(pwfile);
+        pwout.println(password);
+        pwout.close();
+    }
+}
+```
+
+Your application may utilize SAIFE’s zero-knowledge password reset (ZKPR) capability.  In the event that a user forgets their password for the application, an administrator can allow the user to unlock the endpoint, all without revealing the user’s application password.
+
+Password reset capability is enabled during the provisioning process, when the endpoint generates two ZKPR splits.  One split is stored on the endpoint, while another is encrypted – via the AES algorithm – and then sent to management services via a Continuum server.
```
Comment (matthewwwalters) [inline source/index.md]: "when the endpoint generates two ZKPR splits." This was a little confusing to me.  When a key is split (for zkpr purposes), it is split into two parts (technically, you could split the key into more parts, but that's a digression).  I believe the endpoint generates a ZKPR key which which is then split into two parts.  Is that more clear than saying there are two ZKPR splits?

File: source/index.md [hunk 2]
```diff
@@ -300,7 +300,7 @@ Generating a keypair involves:
...
 	* The distinguished name structure, which is independent of the SAIFE Management Dashboard, only requires a common name.  The common name is like a nickname for the certificate and is helpful for debugging purposes.  There isn’t a strict format for the common name; for example, it can be a person’s name, an ID, or a location.
 * generating entropy; and
-	* An appropriate entropy source, whether user-supplied or timing-based, is necessary to seed the deterministic random bit generator (DRBG) with a sequence of random bytes used to generate the keypair.  An acceptable level of entropy ensures that the private key is sufficiently random (and thus, difficult for a potential attacker to determine).
+	* An appropriate entropy source – typically a local system resource such as the /dev/random special file in Unix-like operating systems or the SecureRandom constructor in Java-like platforms – is necessary to seed the deterministic random bit generator (DRBG) with a sequence of random bytes used to generate the keypair.  An acceptable level of entropy ensures that the private key is sufficiently random (and thus, difficult for a potential attacker to determine).
```
Comment (joekamien) [inline source/index.md]: Why do we remove the mention of user-supplied entropy?

File: source/index.md [hunk 3]
```diff
@@ -1127,13 +1127,529 @@ saife.removeVolume(vol);
...
+
+Your application may utilize SAIFE’s zero-knowledge password reset (ZKPR) capability.  In the event that a user forgets their password for the application, an administrator can allow the user to unlock the endpoint, all without revealing the user’s application password.
+
+Password reset capability is enabled during the provisioning process, when the endpoint generates two ZKPR splits.  One split is stored on the endpoint, while another is encrypted – via the AES algorithm – and then sent to management services via a Continuum server.
+
+The password reset process typically involves the following steps:
+
+1. The user, attempting to unlock the SAIFE-enabled application, realizes that they do not have the correct password.
+2. The user contacts their administrator using an out-of-band channel (such as a voice call or email), requesting a password reset.
+	* Ideally, an out-of-band authentication mechanism (such as an SMS verification code) should be used in order to minimize the threat of social engineering.
```
Comment (matthewwwalters) [inline source/index.md]: I'm not sure that this solves the out-of-band authentication problem.  If I'm developing a Android app, for example, then my phone number will be the same device that is running the application.  If I'm a bad guy, steal your phone, then request a password reset, you haven't really authenticated me.  All you did is make sure that I am in possession of the device.  

We may want to simply eliminate the example "(such as an SMS verification code)".

File: source/index.md [hunk 4]
```diff
@@ -1127,13 +1127,529 @@ saife.removeVolume(vol);
...
+3. The administrator resets the password for the pertinent certificate, either in the SAIFE Management Dashboard or using the SAIFE Management API directly.
+4. Management services sends a secure password reset message (containing management services’ ZKPR split) to the endpoint through SAIFE’s network.
```
Comment (matthewwwalters) [inline source/index.md]: Suggest rewording to:

"Management services sends a password reset message with the ZKPR split that was previously stored for the associated endpoint through SAIFE's network".

Note that the message is not encrypted and in the clear.  It can't be encrypted because the endpoint cannot unlock to get the private key.  Therefore, the message is not encrypted so we shouldn't call it "secure".  It is hashed for authentication purposes, however.

File: source/index.md [hunk 5]
```diff
@@ -1127,13 +1127,529 @@ saife.removeVolume(vol);
...
+5. The application unlocks itself and prompts the user for a new password.
```
Comment (matthewwwalters) [inline source/index.md]: The SDK actually automatically unlocks itself.  The application is responsible for prompting the user for the password.

File: source/index.md [hunk 6]
```diff
@@ -1127,13 +1127,529 @@ saife.removeVolume(vol);
...
+6. The user enters a new password, allowing for continued normal use of the application.
+7. The application generates two new password reset splits, sending one split – encrypted via the AES algorithm – to management services through SAIFE’s network.
```
Comment (matthewwwalters) [inline source/index.md]: Step #7 actually occurs in step 5.  The new split is generated and sent up to SAIFE Management Services before the application is notified.  The goal is to minimize the amount of time where the app/device could crash.

I would move #7 to #5, and then #5 would go to #6.

File: source/index.md (hunk 2 above)
Comment (jacobchughes) [inline source/index.md]: I'm under the impression that most developers will use local resources for entropy.  SAIFE Mobile uses user-supplied activity, but this might be beyond the scope of work that most developers are comfortable with.  I could very well be wrong.

File: source/index.md (hunk 1 above)
Comment (jacobchughes) [inline source/index.md]: This should be clearer:

"Password reset capability is enabled during the provisioning process, when the endpoint generates a ZKPR key that is split into two halves.  One half is stored on the endpoint, while the other half is encrypted – via the AES algorithm – and then sent to management services via a Continuum server."

File: source/index.md (hunk 3 above)
Comment (jacobchughes) [inline source/index.md]: Fair enough.

File: source/index.md (hunk 4 above)
Comment (jacobchughes) [inline source/index.md]: Sounds good.

File: source/index.md (hunk 2 above)
Comment (joekamien) [inline source/index.md]: I like the change which mentions typical behavior, but I think we should still mention user-supplied entropy. We could add a sentence after "typically..." to say something like "Some developers utilize user-supplied entropy, for example by having users draw randomly on a screen or shake a device."
